rust/src/builder.rs

```diff
@@ -36,13 +37,99 @@ where
     _marker: PhantomData<T>,
 }
 
-impl<T> BufferBuilder<T>
-where
-    T: ArrowPrimitiveType,
-{
+macro_rules! impl_buffer_builder {
```

Comment:
This code duplication is unfortunate. I really hope we can use [specialization](https://github.com/rust-lang/rust/pull/31844) but it's only available in nightly :(